feathr_project/test/test_derived_features.py

```diff
@@ -0,0 +1,187 @@
+
+    # A derived feature
+    derived_feature = DerivedFeature(name="user_embemdding_derived",
+                                        feature_type=FLOAT,
+                                        key=user_key,
+                                        input_features=user_embedding,
+                                        transform="if_else(user_embedding, user_embedding, [])")
+
+    derived_feature_config = """
+    user_embemdding_derived: {
+        key: [user_id]
+        inputs: {
+            user_embedding: {key: [user_id], feature: user_embedding}
+        }
+        definition: "if_else(user_embedding, user_embedding, [])"
+        type: {
+            type: TENSOR
+            tensorCategory: DENSE
+            dimensionType: []
+            valType: FLOAT
+        }
+    }"""
+    assert_config_equals(derived_feature.to_feature_config(), derived_feature_config)
+
+def test_multikey_derived_feature_to_config():
+    """Multikey derived feature config generation should work"""
+    user_key = TypedKey(full_name="mockdata.user", key_column="user_id", key_column_type=ValueType.INT32, description="An user identifier")
+    item_key = TypedKey(full_name="mockdata.item", key_column="item_id", key_column_type=ValueType.INT32, description="An item identifier")
+    
+    user_embedding = Feature(name="user_embedding", feature_type=FLOAT_VECTOR, key=user_key)
+    item_embedding = Feature(name="item_embedding", feature_type=FLOAT_VECTOR, key=item_key)
+
+    # A derived feature
+    user_item_similarity = DerivedFeature(name="user_item_similarity",
+                                        feature_type=FLOAT,
+                                        key=[user_key, item_key],
+                                        input_features=[user_embedding, item_embedding],
+                                        transform="similarity(user_embedding, item_embedding)")
+
+    derived_feature_config = """
+    user_item_similarity: {
+        key: [user_id, item_id]
+        inputs: {
+            user_embedding: {key: [user_id], feature: user_embedding}
+            item_embedding: {key: [item_id], feature: item_embedding}
+        }
+        definition: "similarity(user_embedding,item_embedding)"
+        type: {
+            type: TENSOR
+            tensorCategory: DENSE
+            dimensionType: []
+            valType: FLOAT
+        }
+    }"""
+    assert_config_equals(user_item_similarity.to_feature_config(), derived_feature_config)
+
+
+def test_derived_feature_to_config_with_alias():
+    # More complicated use case, viewer viewee aliasged user key
+    # References the same key feature with different alias
+    user_key = TypedKey(full_name="mockdata.user", key_column="user_id", key_column_type=ValueType.INT32, description="An user identifier")
+    user_embedding = Feature(name="user_embedding", key=user_key, feature_type=FLOAT_VECTOR)
+    viewer_viewee_distance = DerivedFeature(name="viewer_viewee_distance",
```

Comment:
I was thinking something like
```
class FeatureQuery:
  
    def __init__(self, feature: Feature, key: Optional[Union[TypedKey, List[TypedKey]]] = None, alias: Optional[str] = None) -> None:
        self.feature = feature
        self.key = key
        self.alias = alias

viewer_key = TypedKey(..., key_column="viewer_id")
viewee_key = TypedKey(..., key_column="viewee_id")
viewer_viewee_distance = DerivedFeature(name="viewer_viewee_distance",
                    key=[viewer_key, viewee_key],
                    feature_type=FLOAT,
                    input_features=[FeatureQuery(feature=user_embedding, key=viewer_key, alias="viewer_embedding"),
                                    FeatureQuery(feature=user_embedding, key=viewee_key, alias="viewee_embedding")],
                    transform="distance(viewer_embedding, viewee_embedding)")
```

FeatureQuery is query on a per feature level
You can use FeatureQuery as an input to a derived feature
The key tag notation in derived features is trying to express a FeatureQuery essentially?
Makes it also explicit that the same key object is used for the DerivedFeature key, as well as for the query for the input feature. I think it also reflects more clearly on what the HOCON config version of key-tagged features are
```
      inputs: {
            viewer_embedding: {key: [viewer], feature: user_embedding}
            viewee_embedding: {key: [viewee], feature: user_embedding}
        }
```
This is a feature query, using a key query (e.g. viewer or viewee) to retrieve some feature by name (e.g. user_embedding) and doing an optional renaming .